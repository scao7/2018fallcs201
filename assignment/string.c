#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "string.h"

struct STRING
    {
    char *value;
    };

STRING *
newSTRING(char *x)
    {
    STRING *p = malloc(sizeof(STRING));
    assert(p != 0);
    p->value = x;
    return p;
    }

char * 
getSTRING(STRING *v)
    {
    return v->value;
    }

char * 
setSTRING(STRING *v,char *x)
    {
    char *old = v->value;
    v->value = x;
    return old;
    }

void 
displaySTRING(void *v,FILE *fp)
    {
    fprintf(fp,"%s",getSTRING((STRING *) v));
    }

/*
char * 
compareSTRING(void *v,void *w)
    {
    return getSTRING(v) - getSTRING(w);
    }
*/

void
freeSTRING(void *v)
    {
    free((STRING *) v);
    }