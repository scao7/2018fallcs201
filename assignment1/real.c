#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL
    {
    double value;
    };

REAL *
newREAL(double x)
    {
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
    }

double 
getREAL(REAL *v)
    {
    return v->value;
    }

double 
setREAL(REAL *v,double x)
    {
    double old = v->value;
    v->value = x;
    return old;
    }

void 
displayREAL(void *v,FILE *fp)
    {
    fprintf(fp,"%f",getREAL((REAL *) v));
    }

int 
compareREAL(void *v,void *w)
    {
		double a= getREAL(v);
		double b= getREAL(w);
		if(a==b) return 0;
		else if (a<b) return -1;
		else return 1;
    }
int 
rcompareREAL(void *v, void* w)
	{
		double a= getREAL(v);
		double b= getREAL(w);
		if(a==b) return 0;
		else if (a>b) return -1;
		else return 1;
	}
void
freeREAL(void *v)
    {
    free((REAL *) v);
    }
