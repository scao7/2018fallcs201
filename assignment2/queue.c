/*
written by Shengting Cao 
University of Alabama 
2018 SPRING 
CS201 Algorithm and Data Structure
queue
*/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "sll.h"
#include "queue.h"

typedef struct queue
{
    SLL *list;
    void (*display)(void *, FILE *);
    void (*free) (void *);
}QUEUE;
//////////////////////////////////
	
QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *))
{
   QUEUE *q;
   q = malloc(sizeof(QUEUE));
   assert(q != 0);
   q->display = d;
   q->free = f;
   q->list = newSLL(d,f);
   return q;
}

void enqueue(QUEUE *q ,void *value) {
    if (q ->list == NULL)
        insertSLL(q ->list,0,value);
    else
        insertSLL(q ->list,sizeSLL(q->list),value);
}

void *dequeue(QUEUE *q) {
	assert(sizeSLL(q->list) >0);
    return removeSLL(q->list,0);
}

void *peekQUEUE(QUEUE *q) {
	assert(sizeSLL(q->list) >0);
    return getSLL(q->list,0);
}

int sizeQUEUE(QUEUE *q) {
    return sizeSLL(q->list);
}

void displayQUEUE(QUEUE *q,FILE *fp) 
{
	fprintf(fp,"<");
	for(int i=0;i<sizeQUEUE(q);i++){
		q->display(getSLL(q->list,i),fp);
		if(i != sizeQUEUE(q)-1)
			fprintf(fp,",");
	}
	fprintf(fp,">");
}

void displayQUEUEdebug(QUEUE *q,FILE *fp)
{
	displaySLLdebug(q->list,fp);
}

void freeQUEUE(QUEUE *q)
{
	freeSLL(q->list);
	free(q);
}
