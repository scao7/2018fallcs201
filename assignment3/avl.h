/*** AVL binary search tree class ***/

    #ifndef __AVL_INCLUDED__
    #define __AVL_INCLUDED__

    #include <stdio.h>

    typedef struct avl AVL;

    extern AVL *newAVL(
        void (*)(void *,FILE *),           //display
        int (*)(void *,void *),            //comparator
        void (*)(void *));                 //freeing function
    extern void insertAVL(AVL *,void *);
    extern int findAVLcount(AVL *,void *);
    extern void *findAVL(AVL *,void *);
    extern void *deleteAVL(AVL *,void *);
    extern int sizeAVL(AVL *);
    extern int duplicatesAVL(AVL *);
    extern void statisticsAVL(AVL *,FILE *);
    extern void displayAVL(AVL *,FILE *);
    extern void displayAVLdebug(AVL *,FILE *);
    extern void freeAVL(AVL *);

    #endif
