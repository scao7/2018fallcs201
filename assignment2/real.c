#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL
    {
    double value;
    };

REAL *
newREAL(double x)
    {
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
    }

double 
getREAL(REAL *v)
    {
    return v->value;
    }

double 
setREAL(REAL *v,double x)
    {
    double old = v->value;
    v->value = x;
    return old;
    }

void 
displayREAL(void *v,FILE *fp)
    {
    fprintf(fp,"%f",getREAL((REAL *) v));
    }

int 
compareREAL(void *v,void *w)
    {
    return getREAL(v) - getREAL(w);
    }
int 
rcompareREAL(void *v, void* w)
	{
	return getREAL(w)-getREAL(v);
	}
void
freeREAL(void *v)
    {
    free((REAL *) v);
    }
