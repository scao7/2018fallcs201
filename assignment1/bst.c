/*
written by Shengting Cao 
University of Alabama 
2018 SPRING 
CS201 Algorithm and Data Structure
binary search tree
*/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "queue.h"
#include "bst.h"

typedef struct bstnode{
	void* value;
	struct bstnode* left;
	struct bstnode* right;
	struct bstnode* parent;
}BSTNODE;
/////////////////////////////////////

BSTNODE *newBSTNODE(void *v){
	BSTNODE *items = malloc(sizeof(BSTNODE));
	assert(items !=0);
	items -> value = v;
	items -> left = 0;
	items -> right = 0;
	items -> parent = 0;
	return items;
}

void *getBSTNODEvalue(BSTNODE *n){
	return n->value;
}

void setBSTNODEvalue(BSTNODE *n, void *value){
	n->value = value;
	return;
}

BSTNODE *getBSTNODEleft(BSTNODE *n){
	return n->left;
}

void setBSTNODEleft(BSTNODE *n,BSTNODE *replacement){
	n->left = replacement;
	return;
}
BSTNODE *getBSTNODEright(BSTNODE *n){
	return n-> right;
}
void setBSTNODEright(BSTNODE *n,BSTNODE *replacement){
	n-> right = replacement;
	return;
}
BSTNODE *getBSTNODEparent(BSTNODE *n){
	return n->parent;
}
void setBSTNODEparent(BSTNODE *n,BSTNODE *replacement){
	n->parent=replacement;
	return;
};
void freeBSTNODE(BSTNODE *n,void (*f)(void *)){
	if(f!=NULL)
		f(n->value);
		free(n);
	return;
}
////////////////////////////////////////////////////////////////
typedef struct bst{
	struct bstnode *root;
	int size;
	void(*display)(void *, FILE *);
	int (*comparator)(void *, void *);
	void(*swapper)(BSTNODE *, BSTNODE *);
	void(*free)(void *);
}BST;
//swapper function
static void sw(BSTNODE* a, BSTNODE* b){
	void* temp = getBSTNODEvalue(a);
	setBSTNODEvalue(a,getBSTNODEvalue(b));
	setBSTNODEvalue(b,temp);
}
BST *newBST(void (*d)(void *, FILE *), int (*c)(void *, void *), void(*s)(BSTNODE *, BSTNODE*),void (*f)(void *)){
	BST *tree = malloc(sizeof(BST));
	assert(tree != 0);
	tree -> size = 0;
	tree -> root = 0;
	tree -> display = d;
	tree -> comparator = c;
	tree -> swapper = s;
	if (s == NULL){
		tree -> swapper =sw; //if swapper is null using own swapper function
	}
	tree -> free = f;
	return tree;
}

BSTNODE *getBSTroot(BST *t){
	return t->root;
}
void setBSTroot(BST *t,BSTNODE *replacement){
	t->root = replacement;
	return;
}

void setBSTsize(BST *t,int s){
	t->size =s;
	return;
}

BSTNODE *insertBST(BST *t,void *value){
	BSTNODE* newNode = newBSTNODE(value);
	BSTNODE* check = NULL; // check parent;
	BSTNODE* search = getBSTroot(t);
	while(search != NULL){
		check = search;
		if (t->comparator(getBSTNODEvalue(newNode),getBSTNODEvalue(search))<0)
			search = getBSTNODEleft(search);
		else
			search = getBSTNODEright(search);
	}
	setBSTNODEparent(newNode,check);
	if(check == NULL)//empty tree
		setBSTroot(t,newNode);
	else if(t->comparator(getBSTNODEvalue(newNode), getBSTNODEvalue(check))<0)
		setBSTNODEleft(check, newNode);
	else
		setBSTNODEright(check, newNode);

	setBSTsize(t,sizeBST(t)+1);
	return newNode;
}

BSTNODE *findBST(BST *t,void *value){
	BSTNODE* search = getBSTroot(t);
	while(search != NULL ){
		if(t->comparator(value,getBSTNODEvalue(search))>0)
			search = getBSTNODEright(search);
		else if(t->comparator(value,getBSTNODEvalue(search))<0)
			search = getBSTNODEleft(search);
		else
			return search;
	}
	return NULL;
}

BSTNODE *deleteBST(BST *t,void *value){
	assert(t!=NULL);
	BSTNODE* temp=findBST(t,value);
	if(temp == NULL)
		return NULL;
	temp = swapToLeafBST(t, temp);
	pruneLeafBST(t,temp);
	setBSTsize(t,sizeBST(t)-1);
	return temp;
}
//help fuction to find successor
static BSTNODE* SUCCESSOR(BSTNODE* node){
	assert(node!=NULL);
	BSTNODE* search = node;
	if(getBSTNODEright(search)!=NULL){
		search = getBSTNODEright(search);
	}
	else {
		return NULL;
	}
	while(getBSTNODEleft(search)!= NULL){
		search = getBSTNODEleft(search);
	}
	return search;
}

static BSTNODE* PRECESSOR(BSTNODE* node){
	assert(node!=NULL);
	BSTNODE* search = node;
	if(getBSTNODEleft(search)!=NULL){
		search = getBSTNODEleft(search);
	}
	else {
		return NULL;
	}
	while(getBSTNODEright(search)!= NULL){
		search = getBSTNODEright(search);
	}
	return search;
}
BSTNODE *swapToLeafBST(BST *t,BSTNODE *node){
	if(node == NULL || (getBSTNODEleft(node)== NULL && getBSTNODEright(node)== NULL))
		return node;
	else{
		if(SUCCESSOR(node)){
			t->swapper(node, SUCCESSOR(node));
			return swapToLeafBST(t, SUCCESSOR(node));
		}
		else if(PRECESSOR(node)){
			t->swapper(node, PRECESSOR(node));
			return swapToLeafBST(t, PRECESSOR(node));
		}
		else
			return node;
	}
}

void pruneLeafBST(BST *t,BSTNODE *leaf){
	assert(sizeBST(t) != 0);
	assert(leaf!= NULL);
	BSTNODE* temp = getBSTNODEparent(leaf);
	if(getBSTNODEparent(leaf)==NULL){
		setBSTroot(t,NULL);

	}

	else if (getBSTNODEleft(temp) == leaf){
		setBSTNODEleft(temp, NULL);

	}
	else{
		setBSTNODEright(temp, NULL);
	}
	//setBSTsize(t,sizeBST(t)-1); 
//The pruneLeafBST should not decrement the size
// of the BST. Only the deleteBST method should decrement the count.
	setBSTNODEparent(leaf,NULL);
}
int sizeBST(BST *t){
	return t->size;
}
//help function for max depth
static int maxDepth(BSTNODE* node)
{
   if (node==NULL)
       return -1;
   else
   {
       /* compute the depth of each subtree */
       int lDepth = maxDepth(getBSTNODEleft(node));
       int rDepth = maxDepth(getBSTNODEright(node));

       /* use the larger one */
       if (lDepth > rDepth)
           return(lDepth+1);
       else
		   return(rDepth+1);
   }
}
//get min depth help function
static int minDepth(BSTNODE* node)
{
   if (node==NULL)
       return -1;
   else
   {
       /* compute the depth of each subtree */
       int lDepth = minDepth(node->left);
       int rDepth = minDepth(node->right);

       /* use the small one */
       if (lDepth < rDepth)
           return(lDepth+1);
       else
		   return(rDepth+1);
   }
}

void statisticsBST(BST *t,FILE *fp){
	fprintf(fp, "Nodes: %d\n", t->size);
	fprintf(fp,"Minimum depth: %d\n",minDepth(getBSTroot(t)));
	fprintf(fp,"Maximum depth: %d\n",maxDepth(getBSTroot(t)));
}
//help function for preorder
static void printPreorder(BST* t,BSTNODE* node,FILE* fp)
{
	assert(t!=NULL);
     if (node == NULL){
		  return;
	 }

     /* first print data of node */
	 if(node != getBSTroot(t))
		fprintf(fp," ");
	 fprintf(fp,"[");
	 if(getBSTNODEvalue(node )!=NULL)
     t->display(getBSTNODEvalue(node), fp);
     /* then recur on left subtree */

     printPreorder(t,getBSTNODEleft(node),fp);

     /* now recur on right subtree */
     printPreorder(t,getBSTNODEright(node),fp);
	 fprintf(fp,"]");
}

void displayBST(BST *t,FILE *fp){
	if(getBSTroot(t)==NULL)
		fprintf(fp,"[]");
	printPreorder(t, getBSTroot(t),fp);
}

void displayBSTdebug(BST *t,FILE *fp){
	if(getBSTroot(t)== NULL) return;
	QUEUE* q= newQUEUE(t->display,t->free);
	enqueue(q,getBSTroot(t));//enque node
	BSTNODE* search=getBSTroot(t);
	while(1){
		int nodeCount = sizeQUEUE(q);
		if(nodeCount == 0)
			break;
		while (nodeCount > 0){
			search = peekQUEUE(q);

			t->display(getBSTNODEvalue(search),fp);
			if(nodeCount != 1)
				fprintf(fp," ");
			dequeue(q);
			if(getBSTNODEleft(search))
				enqueue(q,getBSTNODEleft(search));

			if(getBSTNODEright(search))
				enqueue(q,getBSTNODEright(search));

			nodeCount--;
		}

		fprintf(fp,"\n");
	}
	freeQUEUE(q);

}
//healp function for free
static void deallocate (BST* t, BSTNODE* node){
    //do nothing if passed a non-existent node
    if (node == NULL)
        return;

    //now onto the recursion
    deallocate(t,getBSTNODEleft(node));
    deallocate(t,getBSTNODEright(node));

    freeBSTNODE(node, t->free);
}
void freeBST(BST *t){
	deallocate(t,getBSTroot(t));
	free(t);

}
