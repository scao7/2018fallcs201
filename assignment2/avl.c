/*
written by Shengting Cao 
University of Alabama 
2018 SPRING 
CS201 Algorithem and Data Structure
AVL tree 
*/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "bst.h"
#include "avl.h"
typedef struct aval{
	void* value;
	int freq;
	int height;
	int balance;
	void (*display)(void *,FILE *);
	int (*comparator)(void *,void *);
	void (*free)(void *);
}AVAL;

AVAL* newAVAL(void* value, int freq,void (*d)(void *,FILE *),int (*c)(void *,void *),void (*f)(void *)){
	AVAL* x = malloc(sizeof(AVAL));
	assert(x!=0);
	x->height = 1;
	x->balance = 0;
	x ->value= value;
	x ->freq = freq;
	x -> display = d;
	x -> free = f;
	x -> comparator = c;
	return x;
}


void* getAVALvalue(AVAL* x){	// get the original value in avl value
	return x->value;
}


int getAVALfreq(AVAL* x){	//get frequency in avl value
	return x->freq;
}

int getAVALbalance(AVAL* x){
	return x->balance;
}
int getAVALheight(AVAL* x){
	return x->height;
}

void increaseAFreq(AVAL* x){	// increase the frequency
	x-> freq ++;
}

void decreaseAFreq(AVAL* x){
	x-> freq --;
}

int ac(void* G1, void* G2){	//they are void type but we pass avl value to it
	AVAL* gv1 = G1;
	AVAL* gv2 = G2;

	return gv1 -> comparator(getAVALvalue(gv1),getAVALvalue(gv2));
}

void ad(void* G1,FILE* fp){		//G1 is void type but pass avl value to it
	AVAL* gv = G1;
	gv->display(getAVALvalue(gv),fp);
	//printf("height is %d", gv->height );
	//printf("balance value here is %d\n",gv ->balance);
	if(gv->freq !=1){
	fprintf(fp,"[%d]",gv->freq);
	}
	if(getAVALbalance(gv) == 1 )
		fprintf(fp,"+");
	if(getAVALbalance(gv)== -1)
		fprintf(fp,"-");
}

void af(void *n){	//n is void type but pass avl value to it
	AVAL* gv = n;
	if(gv->free!=NULL)
		gv->free(gv->value);
	free(n);
}
static void asw(BSTNODE *a,BSTNODE *b)
        {
        AVAL *ta = getBSTNODEvalue(a);
        AVAL *tb = getBSTNODEvalue(b);


        void *vtemp = ta->value;
        ta->value = tb->value;
        tb->value = vtemp;


        int ctemp = ta->freq;
        ta->freq = tb->freq;
        tb->freq = ctemp;

}

typedef struct avl{
  BST* tree;
  int size;
  void(*display)(void *, FILE *);
  int (*comparator)(void *, void *);
  void(*free)(void *);
} AVL;


AVL *newAVL(void (*d)(void *,FILE *),int (*c)(void *,void *),void (*f)(void *)){
  AVL * g = malloc(sizeof(AVL));
  assert(g != 0);
  g -> tree = newBST(ad,ac,asw,af);
  g -> size = 0;
  g -> display = d;
  g -> comparator = c;
  g -> free = f;
  return g;
}
//set balance
void setBalance(BSTNODE* x){
	BSTNODE* left = getBSTNODEleft(x);
	BSTNODE* right = getBSTNODEright(x);
	AVAL* v = getBSTNODEvalue(x);
	int leftheight;
	int rightheight;
	if(getBSTNODEleft(x) == NULL)
		leftheight =0;
	else{
		leftheight = getAVALheight(getBSTNODEvalue(left));
	}

	if(getBSTNODEright(x) == NULL)
		rightheight =0;
	else{
		rightheight = getAVALheight(getBSTNODEvalue(right));
	}

	if(leftheight >= rightheight)
		v->height = leftheight+1;
	else
		v->height = rightheight+1;
	v -> balance = leftheight - rightheight;
}

BSTNODE* favorite(BSTNODE* p){//pass a parent node return the favorite child
	AVAL* v = getBSTNODEvalue(p);
	if(getAVALbalance(v) > 0)
		return getBSTNODEleft(p);
	else if(getAVALbalance(v) < 0)
		return getBSTNODEright(p);
	else
		return NULL;
}

BSTNODE* sibling(BSTNODE* c){
	BSTNODE* p = getBSTNODEparent(c);
	assert(p);
	if(c == getBSTNODEleft(p))
		return getBSTNODEright(p);

	if(c == getBSTNODEright(p))
		return getBSTNODEleft(p);

	return NULL;
}
//check linear
int linear(BSTNODE* gp, BSTNODE* p, BSTNODE* c){
	return (getBSTNODEleft(gp)== p && getBSTNODEleft(p) == c)|| (getBSTNODEright(gp) == p && getBSTNODEright(p) == c);
}
//left rotation
void leftRotate(BSTNODE* p, BSTNODE* x){
	setBSTNODEright(p,getBSTNODEleft(x));	//p->right = x->left;
    if(getBSTNODEleft(x) !=0) setBSTNODEparent(getBSTNODEleft(x),p);   //if (x->left != 0) x->left->parent = p;
	setBSTNODEleft(x,p);//x->left = p;
    setBSTNODEparent(p,x);//p->parent = x;*/
}
//right rotation
void rightRotate(BSTNODE* p, BSTNODE* x){
	setBSTNODEleft(p,getBSTNODEright(x));//p->left = x->right;
    if(getBSTNODEright(x) !=0) setBSTNODEparent(getBSTNODEright(x),p);//   if (x->right != 0) x->right->parent = p;
    setBSTNODEright(x,p); //   x->right = p;
    setBSTNODEparent(p,x);//    p->parent = x;
}
void rotate(BST* t,BSTNODE* a, BSTNODE* b){//rotate a to b
	//handle grandparent
	BSTNODE* grandparent = getBSTNODEparent(b);
	if(grandparent == NULL){
		setBSTNODEparent(a,NULL);
		setBSTroot(t,a);
	}
	else if(getBSTNODEleft(grandparent)== b){
		setBSTNODEleft(grandparent,a);
		setBSTNODEparent(a,grandparent);
	}
	else{
		setBSTNODEright(grandparent,a);
		setBSTNODEparent(a,grandparent);
	}

	//rotate
	if(getBSTNODEleft(b) == a){//if a is left child do right rotation
		rightRotate(b,a);
	}
	else if(getBSTNODEright(b) == a){//if a is right child do left rotation
		leftRotate(b,a);
	}
}
//insert fix
static void insertionFixup(BST* t, BSTNODE* x){
	while(1){
		if(getBSTNODEparent(x) == NULL) //if v is the root
			break;
		else if(sibling(x)&&favorite(getBSTNODEparent(x)) == sibling(x)){
			setBalance(getBSTNODEparent(x));
			//printf("get the p's balance factors is %d\n", getAVALbalance(getBSTNODEvalue(p)));
			break;
		}
		else if(favorite(getBSTNODEparent(x)) == NULL){
			setBalance(getBSTNODEparent(x));
			//printf("get the p's balance factors is %d\n", getAVALbalance(getBSTNODEvalue(p)));
			x = getBSTNODEparent(x);
		}
		else{
			BSTNODE* y = favorite(x);
			BSTNODE* p = getBSTNODEparent(x);
			if(y && linear(p,x,y) == 0){
				rotate(t,y,x);
				rotate(t,y,p);
				setBalance(x);
				setBalance(p);
				setBalance(y);
			}
			else{
				rotate(t,x,p);
				setBalance(p);
				setBalance(x);
			}
			//printf("get the p's balance factors is %d\n", getAVALbalance(getBSTNODEvalue(p)));
			break;
		}
	}
}
void insertAVL(AVL *g,void *value){
	AVAL* newValue = newAVAL(value,1,g->display, g->comparator, g->free);
	if(findAVL(g,value) == NULL){
		//set parent of new insert to NULL if it is root;
		BSTNODE* newInsert = insertBST(g->tree, newValue);
		if(newInsert == getBSTroot(g->tree)){
			setBSTNODEparent(newInsert,NULL);
		}
		insertionFixup(g->tree, newInsert);//take newly inserted node
	}
	else{
		increaseAFreq(findAVL(g,value));
		free(newValue);
	}
	g->size ++;
}

int findAVLcount(AVL *g,void *value){
	//pass value
	if(findAVL(g,value) ==NULL)
		return 0;
	return getAVALfreq(findAVL(g,value));
}

void *findAVL(AVL *g,void *value){	// pass value return
	AVAL* newNode = newAVAL(value,1,g->display, g->comparator, g->free);
	void* result;
	BSTNODE* w = findBST(g->tree,newNode);
	if(w){
		result = getBSTNODEvalue(w);
	}
	else{
		result = NULL;
	}
	free(newNode);
	return result;
}

// delete fix
void deleteFixup(BST* t, BSTNODE* x){
	AVAL* v = getBSTNODEvalue(x);
	v->height = 0;
	while(1){
		if(getBSTroot(t) == x){
			//printf("case1\n" );
			break;
		}
		else if(favorite(getBSTNODEparent(x)) == x){
			//printf("case2\n");
			setBalance(getBSTNODEparent(x));
			x = getBSTNODEparent(x);
		}
		else if(favorite(getBSTNODEparent(x)) == NULL){
			//printf("case3\n");
			setBalance(getBSTNODEparent(x));
			break;
		}
		else{
			BSTNODE* p = getBSTNODEparent(x);
			BSTNODE* z= sibling(x);
			BSTNODE* y= favorite(z);
			if(y && linear(p,z,y)==0){
				//printf("case4\n");
				rotate(t,y,z);
				rotate(t,y,p);
				setBalance(p);
				setBalance(z);
				setBalance(y);
				x = y;
			}
			else{
				//printf("case5\n" );
				rotate(t,z,p);
				setBalance(p);
				setBalance(z);
				if(y==NULL)
					break;
				x=z;
			}
		}
	}
}

void *deleteAVL(AVL *g,void *value){
	AVAL* newNode = newAVAL(value,1,g->display, g->comparator, g->free);
	void* result;
	assert(findBST(g->tree,newNode))	;
	BSTNODE* w = findBST(g->tree,newNode);
	AVAL* x = getBSTNODEvalue(w);
	assert(x);
	if(getAVALfreq(x) == 1){
		BSTNODE* temp = swapToLeafBST(g->tree,w);
		deleteFixup(g->tree, temp);
		pruneLeafBST(g->tree,temp);
		setBSTsize(g->tree,sizeBST(g->tree)-1);
		result = getAVALvalue(getBSTNODEvalue(temp));
		freeBSTNODE(temp,g->free);
	}
	else{
		decreaseAFreq(x);
		result = NULL;
	}
	free(newNode);
	
	g->size --;
	return result;
}

int sizeAVL(AVL *g){
  return sizeBST(g->tree);
}

int duplicatesAVL(AVL *g){
  return g->size - sizeBST(g->tree);
}
//help function for max depth
static int maxDepth(BSTNODE* node)
{
   if (node==NULL)
       return -1;
   else
   {
       /* compute the depth of each subtree */
       int lDepth = maxDepth(getBSTNODEleft(node));
       int rDepth = maxDepth(getBSTNODEright(node));

       /* use the larger one */
       if (lDepth > rDepth)
           return(lDepth+1);
       else
		   return(rDepth+1);
   }
}
//get min depth help function
static int minDepth(BSTNODE* node)
{
   if (node==NULL)
       return -1;
   else
   {
       /* compute the depth of each subtree */
       int lDepth = minDepth(getBSTNODEleft(node));
       int rDepth = minDepth(getBSTNODEright(node));

       /* use the small one */
       if (lDepth < rDepth)
      		return(lDepth+1);
       else
		   	return(rDepth+1);
   }
}

void statisticsAVL(AVL *g,FILE *fp){
	BST* t= g->tree;
	fprintf(fp,"Duplicates: %d\n",duplicatesAVL(g));
	fprintf(fp, "Nodes: %d\n",sizeBST(t));
	fprintf(fp,"Minimum depth: %d\n",minDepth(getBSTroot(t)));
	fprintf(fp,"Maximum depth: %d\n",maxDepth(getBSTroot(t)));
}

void displayAVL(AVL *g,FILE *fp){
	displayBSTdecorated(g->tree, fp);
}

void displayAVLdebug(AVL *g,FILE *fp){
	displayBST(g -> tree,fp);
}

void freeAVL(AVL *g){
	freeBST(g->tree);
	free(g);
}
