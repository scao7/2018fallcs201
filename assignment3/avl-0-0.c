#include <stdio.h>
#include <stdlib.h>
#include "integer.h"
#include "real.h"
#include "string.h"
#include "queue.h"
#include "bst.h"
#include "avl.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(34);
    //simple INTEGER test of AVL
    AVL *p = newAVL(displayINTEGER,compareINTEGER,freeINTEGER);
    insertAVL(p,newINTEGER(2));
    insertAVL(p,newINTEGER(3));
    insertAVL(p,newINTEGER(1));
    statisticsAVL(p,stdout);
    printf("AVL:\n");
    displayAVL(p,stdout);
    printf("debug: ");
    displayAVLdebug(p,stdout);
    printf("\n");
    INTEGER *q = newINTEGER(2);
    printf("is ");
    displayINTEGER(q,stdout);
    printf(" present? %s\n",findAVL(p,q) == 0? "no" : "yes");
    INTEGER *r = newINTEGER(3);
    freeINTEGER(deleteAVL(p,r));
    statisticsAVL(p,stdout);
    printf("AVL:\n");
    displayAVL(p,stdout);
    printf("debug: ");
    displayAVLdebug(p,stdout);
    printf("\n");
    printf("is ");
    displayINTEGER(q,stdout);
    printf(" present? %s\n",findAVL(p,q) == 0? "no" : "yes");
    setINTEGER(r,2);
    freeINTEGER(deleteAVL(p,r));
    statisticsAVL(p,stdout);
    printf("AVL:\n");
    displayAVL(p,stdout);
    printf("debug: ");
    displayAVLdebug(p,stdout);
    printf("\n");
    printf("is ");
    displayINTEGER(q,stdout);
    printf(" present? %s\n",findAVL(p,q) == 0? "no" : "yes");
    printf("size: %d\n",sizeAVL(p));
    freeINTEGER(q);
    freeINTEGER(r);
    freeAVL(p);
    return 0;
    }
