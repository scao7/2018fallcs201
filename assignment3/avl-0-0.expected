Duplicates: 0
Nodes: 3
Minimum depth: 1
Maximum depth: 1
AVL:
0: 2(2)X
1: =1(2)L =3(2)R
debug: [2 [1] [3]]
is 2 present? yes
Duplicates: 0
Nodes: 2
Minimum depth: 0
Maximum depth: 1
AVL:
0: 2+(2+)X
1: =1(2+)L
debug: [2+ [1]]
is 2 present? yes
Duplicates: 0
Nodes: 1
Minimum depth: 0
Maximum depth: 0
AVL:
0: =1(1)X
debug: [1]
is 2 present? no
size: 1
